src/code.cloudfoundry.org/cni-wrapper-plugin/netrules/rule_converter.go

```diff
@@ -59,6 +59,22 @@ func (c *RuleConverter) BulkConvert(ruleSpec []Rule, logChainName string, global
 	return iptablesRules
 }
 
+func (c *RuleConverter) DeduplicateRules(iptablesRules []rules.IPTablesRule) []rules.IPTablesRule {
```

Comment:
Is this function going to be called frequently with lots of rules? Is so we should add a benchmark and optimise it to ensure we don't waste resources.

Reply:
Some basic optimisations would be:
#### Don't allocate in a loop
Declare `var key string` in front of the loop and re-use it each run.

#### Allocate with enough head-room
Slices (and maps as well, I think) grow as you write to them. Each time they grow, the entire backing array has to be copied. So allocating them with a big enough backing array should reduce the number of allocations and copies:

```go
keys := make(map[string]struct{}, len(iptablesRules))
dedupedRules := make([]rules.IPTablesRule, 0, len(iptablesRules))
```

If you have a better way of estimating the final size of the two structures, use that. I've also changed the type, IIRC `struct{}` is even smaller then `bool`. You would then use the existence (or absence) of the key to determine whether a rule already exists or not:

```go
// exists?
_, exists := dedupedRules["my-rule"]
// add
dedupedRules["my-rule"] = struct{}{}
```

<hr>

I'd still like to see a [benchmark](https://pkg.go.dev/testing#hdr-Benchmarks) for this. That would allow you to test further optimisations, such as whether re-slicing the original slice is more performant than allocating a new one or if doing a `slices.Equal` is quicker than a `strings.Join`.

Reply:
Resolving this conversation based on the feedback provided below. Thanks for taking the time to disprove my theories!